contracts/tokenomics/generator/src/state.rs

```diff
@@ -218,3 +236,54 @@ pub fn update_proxy_asset(deps: DepsMut, proxy_addr: &Addr) -> StdResult<()> {
 
     Ok(())
 }
+
+/// Calculates emission boost amount for specified user and generator
+///
+/// **b_u = min(0.4 * b_u + 0.6 * S * (w_i / W), b_u)**
+///
+/// - b_u is the amount of LP tokens a user staked in a generator
+///
+/// - S is the virtual total amount of LP tokens staked in a generator: virtual_amount = totalLP * 0.4;
+/// - w_i is a user’s current vxASTRO balance
+/// - W is the total amount of vxASTRO
+pub(crate) fn calculate_virtual_amount(
+    deps: Deps,
+    cfg: &Config,
+    pool: &PoolInfo,
+    user_info: &UserInfoV2,
+    account: &Addr,
+) -> StdResult<Uint128> {
+    let mut user_vp = Uint128::zero();
+    let mut total_vp = Uint128::zero();
+
+    if let Some(voting_escrow) = &cfg.voting_escrow {
+        user_vp = get_voting_power(deps.querier, voting_escrow, account)?;
+        total_vp = get_total_voting_power(deps.querier, voting_escrow)?;
+    }
+
+    let user_virtual_share = user_info.amount.multiply_ratio(4u128, 10u128);
+    let total_virtual_share = pool.total_virtual_supply.multiply_ratio(6u128, 10u128);
+
+    let vx_share_emission = if !total_vp.is_zero() {
+        Decimal::from_ratio(user_vp, total_vp)
+    } else {
+        Decimal::zero()
+    };
+
+    Ok(user_virtual_share.checked_add(vx_share_emission.checked_mul(total_virtual_share)?)?)
```

Comment:
We do not need checked functions here. I already wrote detailed explanations before. Checked functions are expensive because they are computed within *256 types. Do not use checked functions if calculations will never throw Overflow.